extensions/pooling/src/__tests__/acceptance/pooling.acceptance.ts

```diff
@@ -0,0 +1,370 @@
+    expect(res.toJSON()).to.eql({status: 'started', id: 1});
+    await poolService.release(res);
+    expect(res.toJSON()).to.eql({status: 'started', id: 1});
+    await poolService.stop();
+    expect(res.toJSON()).to.eql({status: 'stopped', id: 1});
+  });
+
+  it('releases pooled binding on context.close', async () => {
+    const poolService = await givenPoolServiceForBinding();
+    const reqCtx = new Context(ctx, 'req');
+    const res = await poolService.acquire();
+    reqCtx.once('close', () => {
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      poolService.release(res);
+    });
+    const requestClosed = once(reqCtx, 'close');
+    reqCtx.close();
+    reqCtx.emit('close');
+    await requestClosed;
+    expect(poolService.pool.isBorrowedResource(res)).to.be.false();
+  });
+
+  it('allows injection of pooling service', async () => {
+    await givenPoolService({max: 16});
+    ctx
+      .bind('resources.ExpensiveResource')
+      .toProvider(ExpensiveResourceProvider);
+    const res1: PooledValue<ExpensiveResource> = await ctx.get(
+      'resources.ExpensiveResource',
+    );
+    expect(res1.value.toJSON()).to.eql({status: 'created', id: 1});
+    const res2: PooledValue<ExpensiveResource> = await ctx.get(
+      'resources.ExpensiveResource',
+    );
+    expect(res2.value.toJSON()).to.eql({status: 'created', id: 2});
+    await res1.release();
+    const res3: PooledValue<ExpensiveResource> = await ctx.get(
+      'resources.ExpensiveResource',
+    );
+    expect(res3.value.toJSON()).to.eql({status: 'created', id: 1});
+  });
+
+  function givenContext() {
+    ctx = new Context('test');
+  }
+
+  function givenPoolService(
+    poolOptions?: Options,
+    status = 'created',
+    ctor: Constructor<ExpensiveResource> = ExpensiveResource,
+  ) {
+    setupPoolingService(poolOptions, status, ctor);
+    return ctx.get(POOLING_SERVICE);
+  }
+
+  function givenPoolServiceForBinding() {
+    ExpensiveResource.id = 1;
+    const MY_RESOURCE = BindingKey.create<ExpensiveResource>('my-resource');
+    ctx.bind(MY_RESOURCE).toClass(ExpensiveResource);
+    const factory = createPooledBindingFactory(MY_RESOURCE);
+    const poolBinding = createBindingFromClass(PoolingService, {
+      [ContextTags.KEY]: POOLING_SERVICE,
+    });
+    ctx.add(poolBinding);
+    ctx
+      .configure<PoolingServiceOptions<ExpensiveResource>>(poolBinding.key)
+      .to({
+        factory,
+      });
+    return ctx.get(POOLING_SERVICE);
+  }
+
+  /**
+   * This simulates a resource that is expensive to create/start. For example,
+   * a datasource has overhead to connect to the database. There will be performance
+   * penalty to use `TRANSIENT` scope and creates a new instance per request.
+   * But it is not feasible to be a singleton for some use cases, for example,
+   * each request may have different security contexts.
+   */
+  class ExpensiveResource implements Poolable {
```

Comment:
We should move this and other artifacts to an artifacts dir.

Reply:
We're not reusing the class for multiple test files. It's probably better to keep it inside the test file itself.